index.js

```diff
@@ -7,12 +7,55 @@
 
 'use strict';
 
-module.exports = function dashify(str) {
-  if (typeof str !== 'string') {
+module.exports = dashify
+
+function dashify(string) {
+  if (typeof string !== 'string') {
     throw new TypeError('expected a string');
   }
-  str = str.replace(/([a-z])([A-Z])/g, '$1-$2');
-  str = str.replace(/[ \t\W]/g, '-');
-  str = str.replace(/^-+|-+$/g, '');
-  return str.toLowerCase();
+
+  var sentence = []
+  var needsDash = false
+
+  for(var i=0; i<string.length; i++) {
+    var char = string[i];
+
+    var isNumber = !Number.isNaN(parseFloat(char));
+
+    if (isNumber) {
+      lower = char
+    } else {
+      var upper = char.toUpperCase();
+      var lower = char.toLowerCase();
+      var isLetter = upper != lower;
+    } 
+
+    var isAlphaNumeric = isNumber || isLetter;
+
+    if (isNumber) {
+      var isUpper = false;
+    } else if (isLetter) {
+      var isUpper = (char == upper);
+    } else {
+      var isUpper = false;
+    }
+
+    if (needsDash && isAlphaNumeric) {
+      sentence.push("-");
+      needsDash = false;
+    } else if (isUpper && sentence.length) {
+      sentence.push("-");
+    }
+
+    if (isAlphaNumeric) {
+      if (lower == "*") {
+        throw new Error("daf")
```

Comment:
I'm curious, what does `daf` mean? also, what's the reasoning behind throwing here? `*` is a valid character that should be converted to `-` I believe.